geode-core/src/distributedTest/java/org/apache/geode/security/TestIntegratedSecurityService.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.apache.geode.security;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.shiro.subject.Subject;
+
+import org.apache.geode.internal.classloader.ClassPathLoader;
+import org.apache.geode.internal.security.IntegratedSecurityService;
+import org.apache.geode.internal.security.shiro.SecurityManagerProvider;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+
+/**
+ * IntegratedSecurityService extension that allows for manipulation of behavior for testing
+ * handling of rare occurrences of responses of the security service in a more predictable way.
+ */
+public class TestIntegratedSecurityService extends IntegratedSecurityService {
```

Comment:
I wonder if we need to do all these changes just for testing. Can we just use mocks and make sure the correct exceptions are thrown?

Reply:
I'm not sure a mock can do all that this implementation does. It's used to collect test data which is used for verification.